### PR TITLE
Address Remaining Search Schemes Review

### LINKED ISSUE #589: Search Schemes - Remaining issues
See review Comments in #508 which have Not been addressed yet.

### PR DESCRIPTION
**Pull request by eseiler**
Resolves #589

### CODE REVIEW

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 1]
```diff
@@ -68,8 +65,8 @@ namespace seqan3::detail
  *
  * ### Exceptions
  *
- * Strong exception guarantee if iterating the query does not change its state and if this is also guaranteed when
- * invoking a possible delegate specified in `cfg`; basic exception guarantee otherwise.
+ * Strong exception guarantee if iterating the query does not change its state and if invoking a possible delegate
```
Comment (smehringer) [inline include/seqan3/search/algorithm/detail/search.hpp]: what is a delegate?

File: include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp [hunk 2]
```diff
@@ -47,7 +47,7 @@ namespace seqan3::detail
  */
 
 //!\brief Object storing information for a search (of a search scheme).
-//        Number of blocks have to be known at compile time.
+//!       Number of blocks have to be known at compile time.
```
Comment (smehringer) [inline include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]: Don't structs always need an `\ingroup` statement?
EDIT: I see that it is grouped in Seqach/Algorithm... Do you know how?

Comment (eseiler) [reply include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]:
```
/*!\addtogroup submodule_search_algorithm
```
In line 45

File: include/seqan3/search/algorithm/detail/search.hpp (hunk 1 above)
Comment (eseiler) [reply include/seqan3/search/algorithm/detail/search.hpp]: A functor that is called for every hit.

File: include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp (hunk 2 above)
Comment (smehringer) [reply include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]: Ah ok thanks

File: include/seqan3/search/algorithm/detail/search_trivial.hpp [hunk 3]
```diff
@@ -75,7 +72,7 @@ namespace seqan3::detail
  *
  * ### Exceptions
  *
- * No-throw guarantee if this is also guaranteed when invoking the delegate.
+ * No-throw guarantee if invoking the delegate also guarantees no-throw.
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search_trivial.hpp]: but then I propose to add a noexcept expression that checks if the call to the delegate is also noexcept.

File: include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp [hunk 4]
```diff
@@ -69,7 +69,7 @@ struct search
 };
 
 //!\brief Object storing information for a search (of a search scheme).
-//        Number of blocks do not have to be known at compile time.
+//!       Number of blocks do not have to be known at compile time.
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]: Is this detail of the doxygen entity?

Comment (eseiler) [reply include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]: I'm not really sure what it is (or if it's supposed to be brief or detail), it shows as
```
Object storing information for a search (of a search scheme). Number of blocks have to be known at compile time.
```
in the docs. So I guess it's a brief?

File: include/seqan3/search/algorithm/detail/search_trivial.hpp (hunk 3 above)
Comment (eseiler) [reply include/seqan3/search/algorithm/detail/search_trivial.hpp]: I added `noexcept(noexcept(delegate))` for the two cases

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 5]
```diff
@@ -183,15 +180,12 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
 }
 
 /*!\brief Search a query or a range of queries in an index.
- *
- * \tparam index_t   Must model seqan3::fm_index_concept.
- * \tparam queries_t Must be a std::ranges::RandomAccessRange over the index's alphabet.
- *                   a range of queries must additionally model std::ranges::ForwardRange.
- *
+ * \tparam index_t    Must model seqan3::fm_index_concept.
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]: Change to must satisfy `seqan3::fm_index_concept`

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 6]
```diff
@@ -183,15 +180,12 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
...
+ * \tparam queries_t  Must be a std::ranges::RandomAccessRange over the index's alphabet.
+ *                    a range of queries must additionally model std::ranges::ForwardRange.
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]: must satisfy

File: include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp (hunk 4 above)
Comment (rrahn) [reply include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]: well that's the question. I would rather put it into an attention section?

File: include/seqan3/search/algorithm/detail/search_trivial.hpp [hunk 7]
```diff
@@ -167,12 +164,10 @@ inline bool search_trivial(iterator_t it, query_t & query, typename iterator_t::
...
 
 /*!\brief Searches a query sequence in an index using trivial backtracking.
- *
- * \tparam abort_on_hit If the flag is set, the search algorithm aborts on the first hit.
- * \tparam index_t      Must model seqan3::fm_index_concept.
- * \tparam query_t      Must be a std::ranges::InputRange over the index's alphabet.
- * \tparam delegate_t   Takes `index::iterator_type` as argument.
- *
+ * \tparam abort_on_hit  If the flag is set, the search algorithm aborts on the first hit.
+ * \tparam index_t       Must model seqan3::fm_index_concept.
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search_trivial.hpp]: satisfy

File: include/seqan3/search/algorithm/search.hpp [hunk 8]
```diff
@@ -51,29 +51,26 @@ namespace seqan3
...
 
 /*!\brief Search a query or a range of queries in an index.
- *
- * \tparam index_t   Must model seqan3::fm_index_concept.
- * \tparam queries_t Must be a std::ranges::RandomAccessRange over the index's alphabet.
- *                   a range of queries must additionally model std::ranges::ForwardRange.
- *
- * \todo Update concepts and documentation of `configuration_t` everywhere once it has been refactored by rrahn.
- *
+ * \tparam index_t    Must model seqan3::fm_index_concept.
```
Comment (rrahn) [inline include/seqan3/search/algorithm/search.hpp]: satisfy

File: include/seqan3/search/algorithm/search.hpp [hunk 9]
```diff
@@ -124,14 +121,11 @@ inline auto search(index_t const & index, queries_t && queries, configuration_t
 
 /*!\brief Search a query or a range of queries in an index.
  *        It will not allow for any errors and will output all matches as positions in the text.
- *
- * \tparam index_t   Must model seqan3::fm_index_concept.
- * \tparam queries_t Must be a std::ranges::RandomAccessRange over the index's alphabet.
- *                   a range of queries must additionally model std::ranges::ForwardRange.
- *
+ * \tparam index_t    Must model seqan3::fm_index_concept.
```
Comment (rrahn) [inline include/seqan3/search/algorithm/search.hpp]: satisfy

File: include/seqan3/search/algorithm/detail/search.hpp (hunk 5 above)
Comment (smehringer) [reply include/seqan3/search/algorithm/detail/search.hpp]: I thought we agreed on `model` and not `satisfy` quite a while ago?

Comment (smehringer) [reply include/seqan3/search/algorithm/detail/search.hpp]: I think this is inconsistent throughout the library...

**Pull request merged**